sdk/storage/azblob/internal/exported/structured_message_response_reader.go

```diff
@@ -0,0 +1,191 @@
+//go:build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+package exported
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/internal/structuredmsg"
+)
+
+// StructuredMessageResponseReader wraps a response body and validates structured messages
+// if the x-ms-structured-body header is present. It strips the structured message framing
+// and returns only the validated data to the caller.
+type StructuredMessageResponseReader struct {
+	reader              *structuredmsg.StructuredMessageReader
+	underlyingBody      io.ReadCloser
+	currentSegmentData  []byte
+	currentSegmentIndex int
+	headerRead          bool
+	closed              bool
+}
+
+// NewStructuredMessageResponseReader creates a new reader that validates structured messages
+// if the structuredBodyType header indicates a structured message format.
+// If structuredBodyType is nil or empty, it returns the original body unchanged.
+func NewStructuredMessageResponseReader(body io.ReadCloser, structuredBodyType *string) (io.ReadCloser, error) {
+	if structuredBodyType == nil || *structuredBodyType == "" {
+		// No structured message, return body as-is
+		return body, nil
+	}
+
+	// Check if it's a structured message format we support
+	// Format: "XSM/1.0;CRC64" or "XSM/1.0;properties=crc64"
+	bodyType := strings.ToUpper(*structuredBodyType)
+	if !strings.HasPrefix(bodyType, "XSM/1.0") {
+		return body, nil
+	}
+
+	// Create structured message reader
+	reader := structuredmsg.NewStructuredMessageReader(body)
+
+	return &StructuredMessageResponseReader{
+		reader:             reader,
+		underlyingBody:     body,
+		currentSegmentData: nil,
+		headerRead:         false,
+		closed:             false,
+	}, nil
+}
+
+// Read implements io.Reader. It reads data from structured message segments,
+// validates CRC64, and returns only the data (stripping framing).
+func (r *StructuredMessageResponseReader) Read(p []byte) (int, error) {
+	if r.closed {
+		return 0, io.EOF
+	}
+
+	// Read header on first read
+	if !r.headerRead {
+		_, err := r.reader.ReadHeader()
+		if err != nil {
+			if err == io.EOF {
```

Comment:
We should be using `errors.Is` instead of direct comparison.  You can also simplify some of the nesting.
```go
if errors.Is(err, io.EOF) {
    return 0, err
} else if err != nil {
    ...
}
```